src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs

```diff
@@ -45,49 +44,45 @@ public OpenTelemetryLoggerProvider(IOptionsMonitor<OpenTelemetryLoggerOptions> o
 
         internal OpenTelemetryLoggerProvider(OpenTelemetryLoggerOptions options)
         {
-            this.options = options ?? throw new ArgumentNullException(nameof(options));
-            this.loggers = new Dictionary<string, ILogger>(StringComparer.Ordinal);
-
-            foreach (var processor in options.Processors)
+            if (options == null)
             {
-                this.AddProcessor(processor);
+                throw new ArgumentNullException(nameof(options));
             }
-        }
 
-        internal IExternalScopeProvider ScopeProvider
-        {
-            get
+            foreach (var processor in options.Processors)
             {
-                if (this.scopeProvider == null)
-                {
-                    this.scopeProvider = new LoggerExternalScopeProvider();
-                }
-
-                return this.scopeProvider;
+                this.AddProcessor(processor);
             }
         }
 
         void ISupportExternalScope.SetScopeProvider(IExternalScopeProvider scopeProvider)
         {
-            // TODO: set existing loggers
             this.scopeProvider = scopeProvider;
+
+            foreach (KeyValuePair<string, OpenTelemetryLogger> logger in this.loggers)
+            {
+                logger.Value.ScopeProvider = scopeProvider;
+            }
         }
 
         public ILogger CreateLogger(string categoryName)
         {
-            lock (this.loggers)
+            if (!this.loggers.TryGetValue(categoryName, out var logger))
```

Comment:
Curious - Is this thread safe (e.g. would it be possible to break the internal state of Dictionary if other threads are trying to write to it)? https://stackoverflow.com/questions/6738634/is-this-non-locked-trygetvalue-dictionary-access-thread-safe

Reply:
Thanks for the tip! Sadly I have this pattern in a bunch of places I now need to go fix up 😭 I updated the code to use a `ConcurrentDictionary` instead. Check it out, LMK if it looks better now. I think it's worth avoiding the lock here because a lot of threads could all be asking for `ILogger`s/writing messages at the same time?